### PR TITLE
KIALI-2392 Add validation for authorization configs

### PR DESCRIPTION
**Pull request by burmanm**
** Describe the change **

This PR adds validations for ServiceRole / ServiceRoleBindings

** Issue reference **

KIALI-2392

** Backwards incompatible? **

Requires kiali-ui PR #1057 for results to be visible

** Documentation **

Requires updates to the kiali.io docs as this generates two new error messages

### CODE REVIEW

**Comment by rhqci** [comment]
Jenkins CI: kiali-core-pr-e2e-test [#694](https://paste.fedoraproject.org/paste/8rHXg-nMu1eHAm5fo79mxg)
 :red_circle: PRT Failed, Please Contact QE

**Comment by rhqci** [comment]
Jenkins CI: kiali-core-pr-e2e-test [#695](https://paste.fedoraproject.org/paste/grTKOoanR2aMb10yNaLIHw)
  * :heavy_check_mark: run-kiali-e2e-tests [#[1435]](https://paste.fedoraproject.org/paste/Qgx-CFop~Du-yMEMbsmogg)

**Comment by rhqci** [comment]
Jenkins CI: kiali-core-pr-e2e-test [#712](https://paste.fedoraproject.org/paste/HKnf3TmRuBT8o5D-03JWmg)
  * :heavy_check_mark: run-kiali-e2e-tests [#[1453]](https://paste.fedoraproject.org/paste/AtEa3ub5VxuLcPnRvqxFoA)

**Comment by rhqci** [comment]
Jenkins CI: kiali-core-pr-e2e-test [#713](https://paste.fedoraproject.org/paste/VI8B9-MziQpc8w7PVS5zow)
  * :heavy_check_mark: run-kiali-e2e-tests [#[1454]](https://paste.fedoraproject.org/paste/ncqfPCWuF7b68oOgPlP2Ew)

File: business/checkers/authorization/service_checker.go [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+			for ruleIndex, ruleSpec := range rules {
+				if rule, ok := ruleSpec.(map[string]interface{}); ok {
+					if servicesSpec, found := rule["services"]; found {
+						if service, ok := servicesSpec.([]interface{}); ok {
+							for _, s := range service {
+								if host, ok := s.(string); ok {
+									fqdn := kubernetes.ParseHost(host, sc.ServiceRole.GetObjectMeta().Namespace, sc.ServiceRole.GetObjectMeta().ClusterName)
+									if fqdn.Namespace != sc.ServiceRole.GetObjectMeta().Namespace {
+										validation := models.Build("servicerole.invalid.namespace", fmt.Sprintf("spec/rules[%d]/services", ruleIndex))
+										validations = append(validations, &validation)
```
Comment (xeviknal) [inline business/checkers/authorization/service_checker.go]: I see you don't add the servicerole index to the path. Is there something behind that makes it better not to add it?

File: business/checkers/authorization/service_checker.go [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+									}
+									if !sc.hasMatchingService(fqdn.Service) {
+										validation := models.Build("servicerole.invalid.services", fmt.Sprintf("spec/rules[%d]/services", ruleIndex))
+										validations = append(validations, &validation)
```
Comment (xeviknal) [inline business/checkers/authorization/service_checker.go]: Same here. Adding the index to the path would give more precise info to users.

File: models/istio_validation.go [hunk 3]
```diff
@@ -147,6 +150,14 @@ var checkDescriptors = map[string]IstioCheck{
...
 		Severity: ErrorSeverity,
 	},
+	"servicerole.invalid.services": {
+		Message:  "Unable to find all the defined services",
+		Severity: ErrorSeverity,
+	},
+	"servicerole.invalid.namespace": {
+		Message:  "ServiceRole can only point to current namespace",
+		Severity: ErrorSeverity,
+	},
```
Comment (xeviknal) [inline models/istio_validation.go]: There is missing one servicerolebinding validation definition in here.
I guess that it makes the list show the following:
istio-validation-missing-validation-descriptor
(the last item in the list should show an error)

And, as @hhovsepy mentioned, in the yaml page, we see this:
servicebinding-validation-error

Comment (hhovsepy) [reply models/istio_validation.go]: yes the notification message is missing here

File: business/checkers/authorization/service_checker.go (hunk 1 above)
Comment (burmanm) [reply business/checkers/authorization/service_checker.go]: Yes, the format. Multiple services are done in the yaml like this: ``services: ["service1", "service2"]``. We don't support inline paths?

File: business/checkers/authorization/service_binding_checker_test.go [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+import (
+	"testing"
+
+	"github.com/kiali/kiali/config"
+	"github.com/kiali/kiali/kubernetes"
+	"github.com/kiali/kiali/models"
+	"github.com/kiali/kiali/tests/data"
+
+	"github.com/stretchr/testify/assert"
```
[the remark this answers is not in the document]
Comment (burmanm) [reply business/checkers/authorization/service_binding_checker_test.go]: Why don't we have automatic tool for this.. grr. This the output that automatic tools currently assign these.

**Comment by hhovsepy** [comment]
Building with  kiali-ui PR #1057 does not load the Istio Config list.
So built only this PR by merging to latest master.
The Validation column is missing values in Istio Config list page:
Screenshot from 2019-03-18 17-46-38

**Comment by burmanm** [comment]
That's because current master-UI does not want to render validation results for those items.

**Comment by rhqci** [comment]
Jenkins CI: kiali-core-pr-e2e-test [#719](https://paste.fedoraproject.org/paste/ssVPa3psm3yztuR90dAheg)
  * :heavy_check_mark: run-kiali-e2e-tests [#[1469]](https://paste.fedoraproject.org/paste/6rsOB-pEYORUWR8AOIMcTA)

**Comment by hhovsepy** [comment]
The YAML page validations are verified:
Screenshot from 2019-03-18 18-16-37
Screenshot from 2019-03-18 18-16-54

**Comment by hhovsepy** [comment]
Screenshot from 2019-03-18 18-18-46

File: business/checkers/authorization/service_checker.go (hunk 1 above)
Comment (xeviknal) [reply business/checkers/authorization/service_checker.go]: oh I see. Yep, we do cover that scenario. The yaml render it converts the inline yaml to a dashed list.

Here an example:
```yaml
apiVersion: "rbac.istio.io/v1alpha1"
kind: ServiceRole
metadata:
  name: products-viewer
  namespace: bookinfo
spec:
  rules:
  - services: ["productpage.istio-system.svc.cluster.local", "productpage", "reviews.othernamespace.svc.cluster.local"]
    methods: ["GET", "HEAD"]
```

The output of the validation is:
Screenshot-2019-3-19 Kiali Console

**Comment by rhqci** [comment]
Jenkins CI: kiali-core-pr-e2e-test [#726](https://paste.fedoraproject.org/paste/1HNN2uG8nJl2ZeaJSu38Yg)
  * :red_circle: run-kiali-e2e-tests [#[1481]](https://paste.fedoraproject.org/paste/Hsq1dZReZYVRtAYfFNvYiw)

**Pull request merged**